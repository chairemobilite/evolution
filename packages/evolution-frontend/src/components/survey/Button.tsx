/*
 * Copyright 2024, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React, { useState, useEffect } from 'react';
import { withTranslation, WithTranslation } from 'react-i18next';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

import { withSurveyContext, WithSurveyContextProps } from '../hoc/WithSurveyContextHoc';
import ConfirmModal from 'chaire-lib-frontend/lib/components/modal/ConfirmModal';
import { ButtonWidgetConfig } from 'evolution-common/lib/services/widgets';
import { WidgetStatus } from '../../services/interviews/interview';
import * as surveyHelper from 'evolution-common/lib/utils/helpers';
import { UserInterviewAttributes } from 'evolution-common/lib/services/interviews/interview';
import { CliUser } from 'chaire-lib-common/lib/services/user/userType';

type ButtonProps = surveyHelper.InterviewUpdateCallbacks & {
    widgetConfig: ButtonWidgetConfig;
    widgetStatus: WidgetStatus;
    loadingState: number;
    interview: UserInterviewAttributes;
    user?: CliUser;
    path: string;
    label?: string;
    section: string;
};

const Button: React.FC<ButtonProps & WithSurveyContextProps & WithTranslation> = (
    props: ButtonProps & WithSurveyContextProps & WithTranslation
) => {
    const [wasMouseDowned, setWasMouseDowned] = useState(false);
    const [loadingState, setLoadingState] = useState(props.loadingState);
    const [modalIsOpened, setModalIsOpened] = useState(false);

    useEffect(() => {
        // wait for any field to blur (unfocus) and save and/or validate, then trigger click:
        if (props.widgetConfig.hideWhenRefreshing) {
            if (props.loadingState === 0 && loadingState > 0 && wasMouseDowned === true) {
                // this is a hack because it will ignore mouseup if triggered and
                // will run clickButton even if the mouse was not upped.
                // but there is no other way to wait for blur to complete with fast clicks
                // (fast clicks with onBlur do not trigger mouseup)
                setWasMouseDowned(false);
                onClickButton();
            }
        }
        setLoadingState(props.loadingState);
    }, [props.loadingState]);

    const onMouseDown = () => {
        setWasMouseDowned(true);
    };

    const onMouseUp = () => {
        if (!props.widgetConfig.hideWhenRefreshing) {
            onClickButton();
        } else {
            setWasMouseDowned(false);
            onClickButton();
        }
    };

    const confirmButton = () => {
        props.widgetConfig.action(
            {
                startUpdateInterview: props.startUpdateInterview,
                startAddGroupedObjects: props.startAddGroupedObjects,
                startRemoveGroupedObjects: props.startRemoveGroupedObjects
            },
            props.interview,
            props.path,
            props.section,
            props.surveyContext.sections,
            props.widgetConfig.saveCallback
        );
    };

    const onClickButton = () => {
        const hasConfirmPopup = props.widgetConfig.confirmPopup !== undefined;
        const confirmPopupConditional = hasConfirmPopup
            ? surveyHelper.parseBoolean(
                  props.widgetConfig.confirmPopup!.conditional,
                  props.interview,
                  props.path,
                  props.user
            )
            : true;

        if (hasConfirmPopup && confirmPopupConditional === true) {
            setModalIsOpened(true);
        } else {
            confirmButton();
        }
    };

    const isLoading = props.widgetConfig.hideWhenRefreshing && props.loadingState > 0;
    const buttonColor = props.widgetConfig.color || 'green';

    if (!props.widgetStatus.isVisible) {
        return null;
    }

    return (
        <div className={props.widgetConfig.align || 'center'}>
            <button
                type="button"
                className={`survey-section__button button ${buttonColor} ${props.widgetConfig.size || 'large'} ${
                    isLoading ? 'disabled' : ''
                }`}
                onMouseDown={onMouseDown}
                onMouseUp={onMouseUp}
                onKeyUp={(e) => {
                    if (e.key === 'enter' || e.key === 'space' || e.keyCode === 13 || e.keyCode === 32) {
                        onClickButton();
                    } else {
                        return;
                    }
                }}
                disabled={isLoading}
            >
                {props.widgetConfig.icon && <FontAwesomeIcon icon={props.widgetConfig.icon} className="faIconLeft" />}
                {surveyHelper.translateString(
                    props.label || props.widgetConfig.label,
                    props.i18n,
                    props.interview,
                    props.path
                )}
            </button>
            {/* confirmPopup below: */}
            {modalIsOpened && props.widgetConfig.confirmPopup && (
                <div>
                    <ConfirmModal
                        isOpen={true}
                        closeModal={() => setModalIsOpened(false)}
                        text={surveyHelper.translateString(
                            props.widgetConfig.confirmPopup.content,
                            props.i18n,
                            props.interview,
                            props.path
                        )}
                        title={
                            surveyHelper.translateString(
                                props.widgetConfig.confirmPopup.title,
                                props.i18n,
                                props.interview,
                                props.path
                            ) || ''
                        }
                        cancelAction={props.widgetConfig.confirmPopup.cancelAction}
                        showCancelButton={props.widgetConfig.confirmPopup.showCancelButton !== false}
                        showConfirmButton={props.widgetConfig.confirmPopup.showConfirmButton !== false}
                        cancelButtonLabel={surveyHelper.translateString(
                            props.widgetConfig.confirmPopup.cancelButtonLabel,
                            props.i18n,
                            props.interview,
                            props.path
                        )}
                        confirmButtonLabel={surveyHelper.translateString(
                            props.widgetConfig.confirmPopup.confirmButtonLabel,
                            props.i18n,
                            props.interview,
                            props.path
                        )}
                        cancelButtonColor={props.widgetConfig.confirmPopup.cancelButtonColor}
                        confirmButtonColor={props.widgetConfig.confirmPopup.confirmButtonColor}
                        containsHtml={props.widgetConfig.containsHtml || false}
                        confirmAction={confirmButton}
                    />
                </div>
            )}
        </div>
    );
};

export default withTranslation()(withSurveyContext(Button));
