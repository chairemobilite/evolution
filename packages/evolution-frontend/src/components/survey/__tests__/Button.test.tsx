/*
 * Copyright 2024, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import _cloneDeep from 'lodash/cloneDeep';
import React from 'react';
import TestRenderer from 'react-test-renderer';
import each from 'jest-each';
import { mount } from 'enzyme';
import { render } from '@testing-library/react';
import { axe, toHaveNoViolations } from 'jest-axe';
expect.extend(toHaveNoViolations);

import { interviewAttributes } from '../../inputs/__tests__/interviewData.test';
import Button from '../Button';
import { WidgetStatus } from '../../../services/interviews/interview';

const userAttributes = {
    id: 1,
    username: 'foo',
    preferences: {  },
    serializedPermissions: [],
    isAuthorized: () => true,
    is_admin: false,
    pages: [],
    showUserInfo: true
};

const commonWidgetConfig = {
    type: 'button' as const,
    label: 'label',
    action: jest.fn()
};

const defaultWidgetStatus: WidgetStatus = {
    path: 'foo',
    isVisible: true,
    modalIsOpen: false,
    isDisabled: false,
    isCollapsed: false,
    isEmpty: false,
    isCustomEmpty: false,
    isValid: true,
    isResponded: true,
    isCustomResponded: false,
    errorMessage: undefined,
    currentUpdateKey: 1,
    value: null
};

const startUpdateInterviewMock = jest.fn();
const startAddGroupedObjectsMock = jest.fn();
const startRemoveGroupedObjectsMock = jest.fn();

beforeEach(() => {
    jest.clearAllMocks();
});

each([
    ['Default values', commonWidgetConfig],
    ['All values set', {
        ...commonWidgetConfig,
        label: 'newLabel',
        hideWhenRefreshing: true,
        containsHtml: true,
        color: 'red',
        iconPath: 'path/to/somewhere',
        align: 'right',
        saveCallback: jest.fn(),
        confirmPopup: {
            title: 'popupTitle',
            content: 'popupContent',
        },
        size: 'small',
        conditional: jest.fn()
    }],
]).describe('Button widget: %s', (_widget, widgetConfig) => {

    test('Render widget', () => {

        const wrapper = TestRenderer.create(
            <Button
                path='home.region'
                widgetConfig={widgetConfig}
                interview={interviewAttributes}
                user={userAttributes}
                widgetStatus={defaultWidgetStatus}
                section={''}
                startUpdateInterview={startUpdateInterviewMock}
                startAddGroupedObjects={startAddGroupedObjectsMock}
                startRemoveGroupedObjects={startRemoveGroupedObjectsMock}
                loadingState={0}
            />
        );
        expect(wrapper).toMatchSnapshot();
    });

    test('Widget accessibility', async () => {
        const { container } = render(
            <Button
                path='home.region'
                widgetConfig={widgetConfig}
                interview={interviewAttributes}
                user={userAttributes}
                widgetStatus={defaultWidgetStatus}
                section={''}
                startUpdateInterview={startUpdateInterviewMock}
                startAddGroupedObjects={startAddGroupedObjectsMock}
                startRemoveGroupedObjects={startRemoveGroupedObjectsMock}
                loadingState={0}
            />
        );
        const results = await axe(container);
        expect(results).toHaveNoViolations();
    });
});

test('Widget invisible, should be null', () => {
    const widgetStatus = _cloneDeep(defaultWidgetStatus);
    widgetStatus.isVisible = false;
    const wrapper = TestRenderer.create(
        <Button
            path='home.region'
            widgetConfig={commonWidgetConfig}
            interview={interviewAttributes}
            user={userAttributes}
            widgetStatus={widgetStatus}
            section={''}
            startUpdateInterview={startUpdateInterviewMock}
            startAddGroupedObjects={startAddGroupedObjectsMock}
            startRemoveGroupedObjects={startRemoveGroupedObjectsMock}
            loadingState={0}
        />
    );
    expect(wrapper).toMatchSnapshot();
});

test('Widget loading, should be disabled', () => {
    // Set hideWhenRefreshing to true and loadingState to 1
    const wrapper = TestRenderer.create(
        <Button
            path='home.region'
            widgetConfig={{
                ...commonWidgetConfig,
                hideWhenRefreshing: true
            }}
            interview={interviewAttributes}
            user={userAttributes}
            widgetStatus={defaultWidgetStatus}
            section={''}
            startUpdateInterview={startUpdateInterviewMock}
            startAddGroupedObjects={startAddGroupedObjectsMock}
            startRemoveGroupedObjects={startRemoveGroupedObjectsMock}
            loadingState={1}
        />
    );
    expect(wrapper).toMatchSnapshot();
});

describe('Button widget: behavioral tests', () => {
    test('Button click, no modal', () => {
        const buttonWidget = mount(<Button
            path='home.region'
            section='test'
            loadingState={0}
            widgetConfig={commonWidgetConfig}
            interview={interviewAttributes}
            user={userAttributes}
            widgetStatus={defaultWidgetStatus}
            startUpdateInterview={startUpdateInterviewMock}
            startAddGroupedObjects={startAddGroupedObjectsMock}
            startRemoveGroupedObjects={startRemoveGroupedObjectsMock}
        />);

        // Find and click (with mousedown/mouseup) on the button itself and make sure the action has been called
        const button = buttonWidget.find('.button');
        button.simulate('mousedown');
        button.simulate('mouseup');

        // The next action should have been called
        buttonWidget.update();
        expect(commonWidgetConfig.action).toHaveBeenCalledTimes(1);
        expect(commonWidgetConfig.action).toHaveBeenCalledWith({
            startUpdateInterview: startUpdateInterviewMock,
            startAddGroupedObjects: startAddGroupedObjectsMock,
            startRemoveGroupedObjects: startRemoveGroupedObjectsMock
        }, interviewAttributes, 'home.region', 'test', {}, undefined);
    });

    test('Button click, with modal, confirmed', () => {
        const widgetConfig = {
            ...commonWidgetConfig,
            confirmPopup: {
                title: 'popupTitle',
                content: 'popupContent',
            },
        }
        const initialProps = {
            path: 'home.region',
            section: 'test',
            loadingState: 0,
            widgetConfig: widgetConfig,
            interview: interviewAttributes,
            user: userAttributes,
            widgetStatus: defaultWidgetStatus,
            startUpdateInterview: startUpdateInterviewMock,
            startAddGroupedObjects: startAddGroupedObjectsMock,
            startRemoveGroupedObjects: startRemoveGroupedObjectsMock,
        }
        const buttonWidget = mount(<Button
            {...initialProps}
        />);

        // Find and click (with mousedown/mouseup) on the button itself and make sure the action has been called
        const button = buttonWidget.find('.button');
        button.simulate('mousedown');
        button.simulate('mouseup');

        // The action should not have been called
        buttonWidget.update();
        expect(commonWidgetConfig.action).not.toHaveBeenCalled();

        // Find and click on the modal's confirm button
        const confirmModal = buttonWidget.find('.react-modal');
        expect(confirmModal).toMatchSnapshot();
        const confirmButton = confirmModal.findWhere(node => node.type() === 'button' && node.text() === 'Confirm');
        confirmButton.first().simulate('click');
        expect(commonWidgetConfig.action).toHaveBeenCalledTimes(1);
        expect(commonWidgetConfig.action).toHaveBeenCalledWith({
            startUpdateInterview: startUpdateInterviewMock,
            startAddGroupedObjects: startAddGroupedObjectsMock,
            startRemoveGroupedObjects: startRemoveGroupedObjectsMock
        }, interviewAttributes, 'home.region', 'test', {}, undefined);
    });

    test('Button click, with modal, cancelled', () => {
        const widgetConfig = {
            ...commonWidgetConfig,
            confirmPopup: {
                title: 'popupTitle',
                content: 'popupContent',
            },
        }
        const initialProps = {
            path: 'home.region',
            section: 'test',
            loadingState: 0,
            widgetConfig: widgetConfig,
            interview: interviewAttributes,
            user: userAttributes,
            widgetStatus: defaultWidgetStatus,
            startUpdateInterview: startUpdateInterviewMock,
            startAddGroupedObjects: startAddGroupedObjectsMock,
            startRemoveGroupedObjects: startRemoveGroupedObjectsMock,
        }
        const buttonWidget = mount(<Button
            {...initialProps}
        />);

        // Find and click (with mousedown/mouseup) on the button itself and make sure the action has been called
        const button = buttonWidget.find('.button');
        button.simulate('mousedown');
        button.simulate('mouseup');

        // The action should not have been called
        buttonWidget.update();
        expect(commonWidgetConfig.action).not.toHaveBeenCalled();

        // Find and click on the modal's cancel button
        const confirmModal = buttonWidget.find('.react-modal');
        expect(confirmModal).toMatchSnapshot();
        const cancelButton = confirmModal.findWhere(node => node.type() === 'button' && node.text() === 'Cancel');
        cancelButton.first().simulate('click');
        expect(commonWidgetConfig.action).not.toHaveBeenCalled();
    });

    test('With loading state and mouse downed and hideWhenRefreshing to true', () => {
        // Create the original widget
        const widgetConfig = {
            ...commonWidgetConfig,
            hideWhenRefreshing: true
        }
        const initialProps = {
            path: 'home.region',
            section: 'test',
            loadingState: 0,
            widgetConfig: widgetConfig,
            interview: interviewAttributes,
            user: userAttributes,
            widgetStatus: defaultWidgetStatus,
            startUpdateInterview: startUpdateInterviewMock,
            startAddGroupedObjects: startAddGroupedObjectsMock,
            startRemoveGroupedObjects: startRemoveGroupedObjectsMock,
        }
        const buttonWidget = mount(<Button
            {...initialProps}
        />);

        // Simulate the mousedown
        const button = buttonWidget.find('.button');
        button.simulate('mousedown');

        // Update the props to set the loading state to 1 in the meantime
        buttonWidget.setProps({
            ...initialProps,
            loadingState: 1
        });
        buttonWidget.update();

        // Decrement the loading state to 0 again, it should trigger the action
        buttonWidget.setProps({
            ...initialProps,
            loadingState: 0
        });
        buttonWidget.update();
        expect(commonWidgetConfig.action).toHaveBeenCalledTimes(1);
        expect(commonWidgetConfig.action).toHaveBeenCalledWith({
            startUpdateInterview: startUpdateInterviewMock,
            startAddGroupedObjects: startAddGroupedObjectsMock,
            startRemoveGroupedObjects: startRemoveGroupedObjectsMock
        }, interviewAttributes, 'home.region', 'test', {}, undefined);
    });
});