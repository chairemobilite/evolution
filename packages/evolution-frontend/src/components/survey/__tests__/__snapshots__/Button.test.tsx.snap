// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button widget: All values set Render widget 1`] = `
<div
  className="right"
>
  <button
    className="survey-section__button button red small "
    disabled={false}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseUp={[Function]}
    type="button"
  >
    newLabel
  </button>
</div>
`;

exports[`Button widget: Default values Render widget 1`] = `
<div
  className="center"
>
  <button
    className="survey-section__button button green large "
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseUp={[Function]}
    type="button"
  >
    label
  </button>
</div>
`;

exports[`Button widget: behavioral tests Button click, with modal, cancelled 1`] = `
[
  <Modal
    ariaHideApp={true}
    bodyOpenClassName="ReactModal__Body--open"
    className="react-modal"
    closeTimeoutMS={0}
    contentElement={[Function]}
    contentLabel="popupTitle"
    isOpen={true}
    onRequestClose={[Function]}
    overlayClassName="react-modal-overlay"
    overlayElement={[Function]}
    parentSelector={[Function]}
    portalClassName="ReactModalPortal"
    preventScroll={false}
    role="dialog"
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={true}
    shouldFocusAfterRender={true}
    shouldReturnFocusAfterClose={true}
  >
    <Portal
      containerInfo={
        <div
          class="ReactModalPortal"
        >
          <div
            class="ReactModal__Overlay ReactModal__Overlay--after-open react-modal-overlay"
          >
            <div
              aria-label="popupTitle"
              aria-modal="true"
              class="ReactModal__Content ReactModal__Content--after-open react-modal"
              role="dialog"
              tabindex="-1"
            >
              <div>
                <div
                  class="center"
                >
                  <markdown
                    remarkplugins="remark-gfm,[object Object]"
                  >
                    popupTitle
                  </markdown>
                </div>
                <markdown
                  class="confirm-popup"
                  remarkplugins="remark-gfm,[object Object]"
                >
                  popupContent
                </markdown>
                <div
                  class="tr__form-buttons-container _center"
                >
                  <div
                    class="center"
                  >
                    <button
                      class="button grey"
                    >
                      Cancel
                    </button>
                  </div>
                  <div
                    class="center"
                  >
                    <button
                      class="button blue"
                    >
                      Confirm
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      }
    >
      <ModalPortal
        ariaHideApp={true}
        bodyOpenClassName="ReactModal__Body--open"
        className="react-modal"
        closeTimeoutMS={0}
        contentElement={[Function]}
        contentLabel="popupTitle"
        defaultStyles={
          {
            "content": {
              "WebkitOverflowScrolling": "touch",
              "background": "#fff",
              "border": "1px solid #ccc",
              "borderRadius": "4px",
              "bottom": "40px",
              "left": "40px",
              "outline": "none",
              "overflow": "auto",
              "padding": "20px",
              "position": "absolute",
              "right": "40px",
              "top": "40px",
            },
            "overlay": {
              "backgroundColor": "rgba(255, 255, 255, 0.75)",
              "bottom": 0,
              "left": 0,
              "position": "fixed",
              "right": 0,
              "top": 0,
            },
          }
        }
        isOpen={true}
        onRequestClose={[Function]}
        overlayClassName="react-modal-overlay"
        overlayElement={[Function]}
        parentSelector={[Function]}
        portalClassName="ReactModalPortal"
        preventScroll={false}
        role="dialog"
        shouldCloseOnEsc={true}
        shouldCloseOnOverlayClick={true}
        shouldFocusAfterRender={true}
        shouldReturnFocusAfterClose={true}
        style={
          {
            "content": {},
            "overlay": {},
          }
        }
      >
        <div
          className="ReactModal__Overlay ReactModal__Overlay--after-open react-modal-overlay"
          onClick={[Function]}
          onMouseDown={[Function]}
          style={{}}
        >
          <div
            aria-label="popupTitle"
            aria-modal={true}
            className="ReactModal__Content ReactModal__Content--after-open react-modal"
            onClick={[Function]}
            onKeyDown={[Function]}
            onMouseDown={[Function]}
            onMouseUp={[Function]}
            role="dialog"
            style={{}}
            tabIndex="-1"
          >
            <div>
              <div
                className="center"
              >
                <Markdown
                  remarkPlugins={
                    [
                      [
                        "remark-gfm",
                        {
                          "singleTilde": false,
                        },
                      ],
                    ]
                  }
                >
                  popupTitle
                </Markdown>
              </div>
              <Markdown
                className="confirm-popup"
                remarkPlugins={
                  [
                    [
                      "remark-gfm",
                      {
                        "singleTilde": false,
                      },
                    ],
                  ]
                }
              >
                popupContent
              </Markdown>
              <div
                className="tr__form-buttons-container _center"
              >
                <div
                  className="center"
                  key="cancel"
                >
                  <button
                    className="button grey"
                    onClick={[Function]}
                  >
                    Cancel
                  </button>
                </div>
                <div
                  className="center"
                  key="confirm"
                >
                  <button
                    className="button blue"
                    onClick={[Function]}
                  >
                    Confirm
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </ModalPortal>
    </Portal>
  </Modal>,
  <ModalPortal
    ariaHideApp={true}
    bodyOpenClassName="ReactModal__Body--open"
    className="react-modal"
    closeTimeoutMS={0}
    contentElement={[Function]}
    contentLabel="popupTitle"
    defaultStyles={
      {
        "content": {
          "WebkitOverflowScrolling": "touch",
          "background": "#fff",
          "border": "1px solid #ccc",
          "borderRadius": "4px",
          "bottom": "40px",
          "left": "40px",
          "outline": "none",
          "overflow": "auto",
          "padding": "20px",
          "position": "absolute",
          "right": "40px",
          "top": "40px",
        },
        "overlay": {
          "backgroundColor": "rgba(255, 255, 255, 0.75)",
          "bottom": 0,
          "left": 0,
          "position": "fixed",
          "right": 0,
          "top": 0,
        },
      }
    }
    isOpen={true}
    onRequestClose={[Function]}
    overlayClassName="react-modal-overlay"
    overlayElement={[Function]}
    parentSelector={[Function]}
    portalClassName="ReactModalPortal"
    preventScroll={false}
    role="dialog"
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={true}
    shouldFocusAfterRender={true}
    shouldReturnFocusAfterClose={true}
    style={
      {
        "content": {},
        "overlay": {},
      }
    }
  >
    <div
      className="ReactModal__Overlay ReactModal__Overlay--after-open react-modal-overlay"
      onClick={[Function]}
      onMouseDown={[Function]}
      style={{}}
    >
      <div
        aria-label="popupTitle"
        aria-modal={true}
        className="ReactModal__Content ReactModal__Content--after-open react-modal"
        onClick={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        onMouseUp={[Function]}
        role="dialog"
        style={{}}
        tabIndex="-1"
      >
        <div>
          <div
            className="center"
          >
            <Markdown
              remarkPlugins={
                [
                  [
                    "remark-gfm",
                    {
                      "singleTilde": false,
                    },
                  ],
                ]
              }
            >
              popupTitle
            </Markdown>
          </div>
          <Markdown
            className="confirm-popup"
            remarkPlugins={
              [
                [
                  "remark-gfm",
                  {
                    "singleTilde": false,
                  },
                ],
              ]
            }
          >
            popupContent
          </Markdown>
          <div
            className="tr__form-buttons-container _center"
          >
            <div
              className="center"
              key="cancel"
            >
              <button
                className="button grey"
                onClick={[Function]}
              >
                Cancel
              </button>
            </div>
            <div
              className="center"
              key="confirm"
            >
              <button
                className="button blue"
                onClick={[Function]}
              >
                Confirm
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </ModalPortal>,
  <div
    aria-label="popupTitle"
    aria-modal={true}
    className="ReactModal__Content ReactModal__Content--after-open react-modal"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    onMouseUp={[Function]}
    role="dialog"
    style={{}}
    tabIndex="-1"
  >
    <div>
      <div
        className="center"
      >
        <Markdown
          remarkPlugins={
            [
              [
                "remark-gfm",
                {
                  "singleTilde": false,
                },
              ],
            ]
          }
        >
          popupTitle
        </Markdown>
      </div>
      <Markdown
        className="confirm-popup"
        remarkPlugins={
          [
            [
              "remark-gfm",
              {
                "singleTilde": false,
              },
            ],
          ]
        }
      >
        popupContent
      </Markdown>
      <div
        className="tr__form-buttons-container _center"
      >
        <div
          className="center"
          key="cancel"
        >
          <button
            className="button grey"
            onClick={[Function]}
          >
            Cancel
          </button>
        </div>
        <div
          className="center"
          key="confirm"
        >
          <button
            className="button blue"
            onClick={[Function]}
          >
            Confirm
          </button>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`Button widget: behavioral tests Button click, with modal, confirmed 1`] = `
[
  <Modal
    ariaHideApp={true}
    bodyOpenClassName="ReactModal__Body--open"
    className="react-modal"
    closeTimeoutMS={0}
    contentElement={[Function]}
    contentLabel="popupTitle"
    isOpen={true}
    onRequestClose={[Function]}
    overlayClassName="react-modal-overlay"
    overlayElement={[Function]}
    parentSelector={[Function]}
    portalClassName="ReactModalPortal"
    preventScroll={false}
    role="dialog"
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={true}
    shouldFocusAfterRender={true}
    shouldReturnFocusAfterClose={true}
  >
    <Portal
      containerInfo={
        <div
          class="ReactModalPortal"
        >
          <div
            class="ReactModal__Overlay ReactModal__Overlay--after-open react-modal-overlay"
          >
            <div
              aria-label="popupTitle"
              aria-modal="true"
              class="ReactModal__Content ReactModal__Content--after-open react-modal"
              role="dialog"
              tabindex="-1"
            >
              <div>
                <div
                  class="center"
                >
                  <markdown
                    remarkplugins="remark-gfm,[object Object]"
                  >
                    popupTitle
                  </markdown>
                </div>
                <markdown
                  class="confirm-popup"
                  remarkplugins="remark-gfm,[object Object]"
                >
                  popupContent
                </markdown>
                <div
                  class="tr__form-buttons-container _center"
                >
                  <div
                    class="center"
                  >
                    <button
                      class="button grey"
                    >
                      Cancel
                    </button>
                  </div>
                  <div
                    class="center"
                  >
                    <button
                      class="button blue"
                    >
                      Confirm
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      }
    >
      <ModalPortal
        ariaHideApp={true}
        bodyOpenClassName="ReactModal__Body--open"
        className="react-modal"
        closeTimeoutMS={0}
        contentElement={[Function]}
        contentLabel="popupTitle"
        defaultStyles={
          {
            "content": {
              "WebkitOverflowScrolling": "touch",
              "background": "#fff",
              "border": "1px solid #ccc",
              "borderRadius": "4px",
              "bottom": "40px",
              "left": "40px",
              "outline": "none",
              "overflow": "auto",
              "padding": "20px",
              "position": "absolute",
              "right": "40px",
              "top": "40px",
            },
            "overlay": {
              "backgroundColor": "rgba(255, 255, 255, 0.75)",
              "bottom": 0,
              "left": 0,
              "position": "fixed",
              "right": 0,
              "top": 0,
            },
          }
        }
        isOpen={true}
        onRequestClose={[Function]}
        overlayClassName="react-modal-overlay"
        overlayElement={[Function]}
        parentSelector={[Function]}
        portalClassName="ReactModalPortal"
        preventScroll={false}
        role="dialog"
        shouldCloseOnEsc={true}
        shouldCloseOnOverlayClick={true}
        shouldFocusAfterRender={true}
        shouldReturnFocusAfterClose={true}
        style={
          {
            "content": {},
            "overlay": {},
          }
        }
      >
        <div
          className="ReactModal__Overlay ReactModal__Overlay--after-open react-modal-overlay"
          onClick={[Function]}
          onMouseDown={[Function]}
          style={{}}
        >
          <div
            aria-label="popupTitle"
            aria-modal={true}
            className="ReactModal__Content ReactModal__Content--after-open react-modal"
            onClick={[Function]}
            onKeyDown={[Function]}
            onMouseDown={[Function]}
            onMouseUp={[Function]}
            role="dialog"
            style={{}}
            tabIndex="-1"
          >
            <div>
              <div
                className="center"
              >
                <Markdown
                  remarkPlugins={
                    [
                      [
                        "remark-gfm",
                        {
                          "singleTilde": false,
                        },
                      ],
                    ]
                  }
                >
                  popupTitle
                </Markdown>
              </div>
              <Markdown
                className="confirm-popup"
                remarkPlugins={
                  [
                    [
                      "remark-gfm",
                      {
                        "singleTilde": false,
                      },
                    ],
                  ]
                }
              >
                popupContent
              </Markdown>
              <div
                className="tr__form-buttons-container _center"
              >
                <div
                  className="center"
                  key="cancel"
                >
                  <button
                    className="button grey"
                    onClick={[Function]}
                  >
                    Cancel
                  </button>
                </div>
                <div
                  className="center"
                  key="confirm"
                >
                  <button
                    className="button blue"
                    onClick={[Function]}
                  >
                    Confirm
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </ModalPortal>
    </Portal>
  </Modal>,
  <ModalPortal
    ariaHideApp={true}
    bodyOpenClassName="ReactModal__Body--open"
    className="react-modal"
    closeTimeoutMS={0}
    contentElement={[Function]}
    contentLabel="popupTitle"
    defaultStyles={
      {
        "content": {
          "WebkitOverflowScrolling": "touch",
          "background": "#fff",
          "border": "1px solid #ccc",
          "borderRadius": "4px",
          "bottom": "40px",
          "left": "40px",
          "outline": "none",
          "overflow": "auto",
          "padding": "20px",
          "position": "absolute",
          "right": "40px",
          "top": "40px",
        },
        "overlay": {
          "backgroundColor": "rgba(255, 255, 255, 0.75)",
          "bottom": 0,
          "left": 0,
          "position": "fixed",
          "right": 0,
          "top": 0,
        },
      }
    }
    isOpen={true}
    onRequestClose={[Function]}
    overlayClassName="react-modal-overlay"
    overlayElement={[Function]}
    parentSelector={[Function]}
    portalClassName="ReactModalPortal"
    preventScroll={false}
    role="dialog"
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={true}
    shouldFocusAfterRender={true}
    shouldReturnFocusAfterClose={true}
    style={
      {
        "content": {},
        "overlay": {},
      }
    }
  >
    <div
      className="ReactModal__Overlay ReactModal__Overlay--after-open react-modal-overlay"
      onClick={[Function]}
      onMouseDown={[Function]}
      style={{}}
    >
      <div
        aria-label="popupTitle"
        aria-modal={true}
        className="ReactModal__Content ReactModal__Content--after-open react-modal"
        onClick={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        onMouseUp={[Function]}
        role="dialog"
        style={{}}
        tabIndex="-1"
      >
        <div>
          <div
            className="center"
          >
            <Markdown
              remarkPlugins={
                [
                  [
                    "remark-gfm",
                    {
                      "singleTilde": false,
                    },
                  ],
                ]
              }
            >
              popupTitle
            </Markdown>
          </div>
          <Markdown
            className="confirm-popup"
            remarkPlugins={
              [
                [
                  "remark-gfm",
                  {
                    "singleTilde": false,
                  },
                ],
              ]
            }
          >
            popupContent
          </Markdown>
          <div
            className="tr__form-buttons-container _center"
          >
            <div
              className="center"
              key="cancel"
            >
              <button
                className="button grey"
                onClick={[Function]}
              >
                Cancel
              </button>
            </div>
            <div
              className="center"
              key="confirm"
            >
              <button
                className="button blue"
                onClick={[Function]}
              >
                Confirm
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </ModalPortal>,
  <div
    aria-label="popupTitle"
    aria-modal={true}
    className="ReactModal__Content ReactModal__Content--after-open react-modal"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    onMouseUp={[Function]}
    role="dialog"
    style={{}}
    tabIndex="-1"
  >
    <div>
      <div
        className="center"
      >
        <Markdown
          remarkPlugins={
            [
              [
                "remark-gfm",
                {
                  "singleTilde": false,
                },
              ],
            ]
          }
        >
          popupTitle
        </Markdown>
      </div>
      <Markdown
        className="confirm-popup"
        remarkPlugins={
          [
            [
              "remark-gfm",
              {
                "singleTilde": false,
              },
            ],
          ]
        }
      >
        popupContent
      </Markdown>
      <div
        className="tr__form-buttons-container _center"
      >
        <div
          className="center"
          key="cancel"
        >
          <button
            className="button grey"
            onClick={[Function]}
          >
            Cancel
          </button>
        </div>
        <div
          className="center"
          key="confirm"
        >
          <button
            className="button blue"
            onClick={[Function]}
          >
            Confirm
          </button>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`Widget invisible, should be null 1`] = `null`;

exports[`Widget loading, should be disabled 1`] = `
<div
  className="center"
>
  <button
    className="survey-section__button button green large disabled"
    disabled={true}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseUp={[Function]}
    type="button"
  >
    label
  </button>
</div>
`;
